### DISCUSSION TITLE
Neoscore 0.1.2 released with improved window and viewport controls

### DISCUSSION THREAD

**Discussion by ajyoon** [DigiScore/neoscore#53] [Announcements]
We're excited to announce the next neoscore release, `0.1.2`! This new version focuses on improving controls of the interactive view and window. You can now:
- Disable mouse viewport scrolling and dragging, including the appearance of scrollbars
- Constrain the window size with minimum and maximum bounds
- Run the interactive view in fullscreen mode
- Programmatically get and set the viewport's position, zoom, and rotation.

Check out [the changelog](https://github.com/DigiScore/neoscore/blob/main/CHANGELOG.md) for more details! And please report any bugs you may stumble into...

**Comment by Xavman42** [comment]
I may have found a bug. `neoscore.set_viewport_scale()` works fine within a refresh_function, but `neoscore.set_viewport_center_pos()` does not animate. Upon closer inspection, `neoscore.set_viewport_center_pos()` seems to have no effect at all. Here's some code that rotates and scales the frame just fine, but does not translate at all:

```
from neoscore.common import *
import numpy as np

neoscore.setup()
text = Text((Mm(50), Mm(50)), None, "text")

def refresh_func(time: float):
    temp_val = Mm(50) + Mm(np.sin(time) * 200)
    neoscore.set_viewport_rotation(np.sin(time/2)*10)
    neoscore.set_viewport_scale((np.sin(3*time/2)+2)*2)
    neoscore.set_viewport_center_pos((temp_val,Unit(0)))

neoscore.show(refresh_func, display_page_geometry=False, auto_viewport_interaction_enabled=False, min_window_size=(1280,720), max_window_size=(1280,720))
```

**Reply by ajyoon** [reply]
This is actually expected behavior, but it's not very intuitive. Per [`set_viewport_center_pos`'s docstring](https://neoscore.org/api/neoscore.core.neoscore.html#neoscore.core.neoscore.set_viewport_center_pos):

> If the given position is such that the requested view goes beyond the document’s bounding rect, the actual position used is adjusted to prevent that. This is caused by an internal limitation and may change in the future.

If you add a second text item at a different location such that the scene bounding rect is wider, you will see it reposition correctly.

```python
text_2 = Text((Mm(300), Mm(300)), None, "text 2")
```

This is due to some annoying Qt implementation details I haven't found a good solution for. The best way to work around this right now is to add an invisible rectangular path which outlines the scene bounding rect:

```python
# Scene bounding rect
Path.rect((Mm(-150), Mm(-100)), None, Mm(300), Mm(200),
          Brush.no_brush(), Pen.no_pen())

text = Text((Mm(50), Mm(50)), None, "text")

# ...
```

I'll add something to this effect to the docs for now

**Reply by ajyoon** [reply]
to work around the issue entirely, you can just make that invisible rect arbitrarily large. Since you aren't using auto viewport interaction, the effect this normally would have on scrollbars won't matter.

**Reply by Xavman42** [reply]
Thank you, this was super helpful! Things are translating just fine now.
